torch_geometric/nn/models/rbcd_attack.py

```diff
@@ -0,0 +1,770 @@
+import abc
+from collections import defaultdict
+from typing import Any, Callable, Dict, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+import torch_sparse
+from torch import Tensor
+from tqdm import tqdm
+
+from torch_geometric.utils import to_undirected
+
+# (predictions, labels, ids/mask) -> Tensor with one element
+LOSS_TYPE = Callable[[Tensor, Tensor, Optional[Tensor]], Tensor]
+
+
+class Attack(torch.nn.Module):
+    """Abstract class for an adversarial attack that perturbs edges:
+
+    aims to answer the question what (small) perturbation changes the model's
+    prediction the most."""
+    @abc.abstractmethod
+    def attack(self, x: Tensor, edge_index: Tensor, labels: Tensor,
+               budget: int, idx_attack: Optional[Tensor] = None,
+               **kwargs) -> Tuple[Tensor, Tensor]:
+        """Attack the predictions for the provided model and graph.
+
+        A subset of predictions may be specified with :attr:`idx_attack`. The
+        attack is allowed to flip (i.e. add or delete) :attr:`budget` edges and
+        will return the strongest perturbation it can find. It returns both the
+        resulting perturbed  :attr:`edge_index` as well as the perturbations.
+
+        Args:
+            x (Tensor): The node feature matrix. We assume `x` to be located
+                on target device.
+            edge_index (LongTensor): The edge indices.
+            labels (Tensor): The labels.
+            budget (int): The number of allowed perturbations (i.e.
+                number of edges that are flipped at most).
+            idx_attack (Tensor, optional): Filter for predictions/labels.
+            **kwargs (optional): Additional arguments passed to the GNN module.
+
+        :rtype: (:class:`Tensor`, :class:`Tensor`)
+        """
+
+        raise NotImplementedError('Abstractmethod needs to be implemented')
+
+    def __call__(self, *args, **kwargs):
+        return self.attack(*args, **kwargs)
+
+
+class RBCDAttack(Attack):
```

Comment:
Just a suggestion: could we implement PRBCG and GRBCD separately and inherit from RBCDAttack? E.g.:
```python
class RBCDAttack(nn.Module):
    ...
class PRBCDAttack(RBCD):
    ...
class GRBCDAttack(RBCD):
  ....
```

Reply:
Yes, I first had it like this but felt like it was (1) unnecessarily inflating the documentation, (2) changing the attack through configuration is convenient, and (3) subclassing it is somehow nicer this way (e.g. see poisoning example).

Of course, the cons are (a) the RBCD class is more complicated this way and (b) there are some attributes that are only used by `PRBCD`.

As I was unsure, I also asked some lab mates about what interface they preferred. The results were roughly 50/50.

So long story short, I am happy to go with the subclass solution, if you think it is a better fit. I do not have a very strong opinion on this tbh.

Reply:
I would vote for the subclass solution as it makes the interfaces clear and understandable for users. 
Alternatively, we can just implement two classes for them:
```python
class GRBCD(nn.Module or Attack):
    ...
class PRBCD(GRBCD):
   ...
```

But this requires a lot of effort for the changes to this version. @rusty1s could you give us some advice?

Reply:
I am happy to implement whatever is preferred. It is also not that much work as the interfaces etc. are all already clear.

Reply:
@EdisonLeeeee I have now switched to the suggested subclass solution. Please let me know if this is the version to continue. I think we should come to a decision on this before I continue any future feedback.

Reply:
LGTM. Will review there again. Thanks for your effort!